pbj-core/pbj-grpc-helidon/docs/design.md

```diff
@@ -0,0 +1,455 @@
+
+Each of the remaining three types of RPC calls involve streaming. Streaming requires the application developer to deal
+with connection termination, cancellation, and error handling. The approach taken by PBJ is similar to that taken in
+`grpc.io`, but instead of using a library-specific `StreamObserver` object, PBJ uses implementations of the new reactive
+`Flow` API in Java 9 and associated interfaces.
+
+The `Flow` API has three primary methods on a `Flow.Subscriber`:
+  - `onNext`: called when a new message is available
+  - `onError`: called when an error occurs
+  - `onComplete`: called when the processing for this subscriber is complete
+
+It also includes some API for managing flow control, hence the name `Flow`. Each subscriber indicates to the publisher
+how many messages it is prepared to receive, and the publisher can then push that many messages (one at a time).
+
+# Implementation Design
+
+This module follows the same general pattern found on the official
+[Helidon gRPC module](https://github.com/helidon-io/helidon/tree/main/webserver/grpc/src/main/java/io/helidon/webserver/grpc).
+Two classes provided by this module are loaded by Helidon:
+- `com.hedera.pbj.grpc.helidon.PbjProtocolConfigProvider`: this is used to provide integration of configuration for
+  this module with the Helidon configuration system.
+- `com.hedera.pbj.grpc.helidon.PbjProtocolProvider`: the primary integration point, handles HTTP2 requests
+
+The `grpc` package of the PBJ runtime contains two classes to support grpc. These are the `ServiceInterface` and a
+utility class called `Pipelines`.
+
+`ServiceInterface` defines common methods related to all service interface implementations. Given any
+`ServiceInterface` implementation, you can introspect the service name, the "full name" (that is, the service name
+plus whatever "package" it is part of, such as would be used as the HTTP path to the service endpoint), and the list
+of methods (along with their names) that make up the service.
+
+Most importantly, the `ServiceInterface` defines the `open` method. This is called at the very beginning of the gRPC
+session creation. The `open` method is responsible for creating a `Flow.Subscriber` that will receive messages from
+the client. The method also receives the method that will be called, any request options, and the `Flow.Subscriber`
+that should be called for `onNext` (sending messages back to the client), `onError` (if there are errors), or
+`onComplete` (if the response is complete).
+
+This method may be called many times concurrently from different threads. Because this gRPC implementation is based
+on Helidon Nima, each request comes in on a single distinct Virtual Thread.
+
+## Pipelines
+
+The `Pipelines` class is a utility class that provides a way to create a "pipeline" of steps involved in processing
+requests and responses. This class is not intended to be used by the application developer. It is used by the code
+generated by the PBJ compiler.
```

Comment:
This might be confused with the platform pipelines.  Do we want to either rename this class or possibly join the two concepts?

Reply:
I'm probably forgetting, but I don't remember pipelines in the platform (are you thinking of the workflows in services, or the wiring framework in platform?). In any event, I think this is OK as they are internal to the PBJ module, and are not public or exposed to anyone.

Reply:
I was referring to platform wiring framework.
I agree that as long as the pipelines remain internal it's probably OK.